mysqltest/docker_test.go

```diff
@@ -0,0 +1,79 @@
+// Copyright 2022 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mysqltest
+
+// This file should only be used outside of Google's build system.
+
+import (
+	"database/sql"
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+)
+
+func TestKillAfter(t *testing.T) {
+	// This "kill after" time was chosen because it's long enough for the MySQL container to start up.
+	// As of 2022-08-31 on MySQL 5.7, it takes 12.5 seconds to start. We add some buffer to leave room
+	// for normal variation between test machines.
+	const (
+		expectedStartupDuration = 13 * time.Second
+		extraBuffer             = 10 * time.Second
+		killAfter               = expectedStartupDuration + extraBuffer
+		killAfterSec            = int(killAfter / time.Second)
+	)
+
+	ci, stopper, err := start(buildConfig(KillAfterSeconds(killAfterSec)))
+	defer stopper()
+	if err != nil {
+		t.Fatal(err)
+	}
+	db := connect(t, ci)
+	defer db.Close()
+
+	if err := db.Ping(); err != nil {
+		t.Fatalf("db.Ping: %v", err)
+	}
+
+	deadline := time.Now().Add(killAfter)
```

Comment:
I'm pretty sure there's a fun potential bug here. If clocks were turned back 1hr (in places where DST exists) while the loop is running, the loop would run for a full hour (because the deadline is actually +1h now). Maybe UTC both to be safe?

Reply:
Also, FWIW, we have a generic retry library that does different back-off algorithms.